File: frontend/src/services/promotionService.ts
import { Promotion } from '@/types';
import api from './api';

export async function createPromotion(payload: {
  vendorId: string;
  name: string;
  description?: string;
  type: 'percentage' | 'free_delivery' | 'bogo';
  value: number;
  minOrderAmount?: number;
  maxDiscount?: number;
  applicableTo?: string;
  applicableIds?: string[];
  isActive?: boolean;
  startsAt: string;
  expiresAt?: string;
  usageLimit?: number;
}): Promise<Promotion> {
  const response = await api.post<Promotion>('/promotions', payload);
  return response.data;
}

export async function getVendorPromotions(vendorId: string): Promise<Promotion[]> {
  const response = await api.get<Promotion[]>(`/vendors/${vendorId}/promotions`);
  return response.data;
}

export async function getActivePromotions(vendorId: string): Promise<Promotion[]> {
  const response = await api.get<Promotion[]>(`/vendors/${vendorId}/promotions/active`);
  return response.data;
}

export async function updatePromotion(
  id: string,
  payload: Partial<{
    name: string;
    description: string;
    type: 'percentage' | 'free_delivery' | 'bogo';
    value: number;
    minOrderAmount: number;
    maxDiscount: number;
    applicableTo: string;
    applicableIds: string[];
    isActive: boolean;
    startsAt: string;
    expiresAt: string;
    usageLimit: number;
  }>,
): Promise<Promotion> {
  const response = await api.patch<Promotion>(`/promotions/${id}`, payload);
  return response.data;
}

export async function togglePromotion(id: string): Promise<Promotion> {
  const response = await api.patch<Promotion>(`/promotions/${id}/toggle`);
  return response.data;
}

export async function deletePromotion(id: string): Promise<void> {
  await api.delete(`/promotions/${id}`);
}

File: flutter_geobite/lib/services/promotion_service.dart
import '../core/api_client.dart';

class PromotionService {
  /// Get all active promotions
  Future<List<Map<String, dynamic>>> getPromotions() async {
    try {
      final response = await apiClient.dio.get('/promotions');
      final data = response.data;
      final List list = data is Map ? (data['data'] ?? []) : data;
      return list.cast<Map<String, dynamic>>();
    } catch (e) {
      throw Exception('Failed to load promotions: $e');
    }
  }

  /// Get promotions for a specific vendor
  Future<List<Map<String, dynamic>>> getVendorPromotions(String vendorId) async {
    try {
      final response = await apiClient.dio.get('/promotions/vendor/$vendorId');
      final data = response.data;
      final List list = data is Map ? (data['data'] ?? []) : data;
      return list.cast<Map<String, dynamic>>();
    } catch (e) {
      throw Exception('Failed to load vendor promotions: $e');
    }
  }

  /// Create a promotion (seller only)
  Future<Map<String, dynamic>> createPromotion(Map<String, dynamic> payload) async {
    try {
      final response = await apiClient.dio.post('/promotions', data: payload);
      return Map<String, dynamic>.from(response.data);
    } catch (e) {
      throw Exception('Failed to create promotion: $e');
    }
  }

  /// Update a promotion (seller only)
  Future<Map<String, dynamic>> updatePromotion(String id, Map<String, dynamic> payload) async {
    try {
      final response = await apiClient.dio.patch('/promotions/$id', data: payload);
      return Map<String, dynamic>.from(response.data);
    } catch (e) {
      throw Exception('Failed to update promotion: $e');
    }
  }

  /// Delete a promotion (seller only)
  Future<void> deletePromotion(String id) async {
    try {
      await apiClient.dio.delete('/promotions/$id');
    } catch (e) {
      throw Exception('Failed to delete promotion: $e');
    }
  }
}

final promotionService = PromotionService();
